src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+ * non-home shard to store borrowed tokens again.
+ *
+ * This is invoked from the home shard when it deems that there
+ * is no longer pressure for tokens.
+ */
+void conn_quota::cancel_reclaim_from(ss::net::inet_address addr) {
+    vlog(rpc::rpclog.trace, "cancel_reclaim_from({})", addr);
+    if (addr == ss::net::inet_address()) {
+        total_remote.reclaim = false;
+    } else {
+        auto found = ip_remote.find(addr);
+        if (found != ip_remote.end()) {
+            found->second.reclaim = false;
+        }
+    }
+}
+
+/**
+ * I am the home shard for this address.  Acquire tokens, reclaiming
+ * if necessary.
+ */
+ss::future<bool> conn_quota::home_get_units(ss::net::inet_address addr) {
+    assert_on_home(addr);
+
+    auto& allowance = get_home_allowance(addr);
+    vlog(rpc::rpclog.trace, "home_get_units({}) allowance={}", addr, allowance);
+
+    bool result = try_get_units(allowance);
+    if (result || allowance.reclaim) {
+        // If we got a token, or we didn't and are already in
+        // reclaim mode, this is the final answer
+        vlog(rpc::rpclog.trace, "home_get_units: fast path got={}", result);
+        co_return result;
+    }
+
+    // We didn't get any units, but there might be some on other cores
+    co_await reclaim_to(allowance, addr, false);
+
+    // This can still fail if there were no reclaimable units, but
+    // we did our best.
+    result = try_get_units(allowance);
+    vlog(
+      rpc::rpclog.trace, "home_get_units: slow (reclaim) path got={}", result);
+    co_return result;
+}
+
+bool conn_quota::try_get_units(home_allowance& allowance) {
+    if (allowance.available) {
+        allowance.available -= 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * Given the home state of an allowance, decide whether it should
+ * leave reclaim (i.e. permit other shards to borrow tokens again)
+ */
+bool conn_quota::should_leave_reclaim(home_allowance& allowance) {
+    return allowance.reclaim
+           // Must be enough tokens for it to be worth borrowing any
+           && allowance.max > ss::smp::count
+           // Must have at least half its tokens free
+           && allowance.available > allowance.max / 2
+           // Must not be in the middle of starting a reclaim
+           && allowance.reclaim_lock.ready();
+}
+
+/**
+ * Broadcast (in the background) to all other shards that they
+ * may clear the reclaim flag.
+ */
+void conn_quota::cancel_reclaim_to(ss::net::inet_address addr) {
+    assert_on_home(addr);
+
+    vlog(rpc::rpclog.trace, "cancel_reclaim_to({})", addr);
+
+    // Guaranteed to have units because of precheck in
+    // should_leave_reclaim
+    ss::semaphore_units units;
+    auto& allowance = get_home_allowance(addr);
+    allowance.reclaim_lock.try_get_units().value();
+    allowance.reclaim = false;
+
+    ssx::spawn_with_gate(_gate, [this, addr, units = std::move(units)] {
```

Comment:
the captured `units` here seem seem to always have a default-initialized value.

Reply:
That's a bug, units was supposed to be the result of taking reclaim_lock.  Fixed.

Reply:
oh crazy. I'm surprised that there isn't some sort of `[[nodiscard]]` happening here `allowance.reclaim_lock.try_get_units().value();` i wonder if it has to do with there being an optional involved?